atomic_reactor/plugins/pre_fetch_sources.py

```diff
@@ -0,0 +1,150 @@
+        workdir = self.workflow.source.get_build_file_path()[1]
+        dest_dir = os.path.join(workdir, self.DOWNLOAD_DIR)
+        downloaded_sources = []
+
+        for url in urls:
+            parsed_url = urlparse(url)
+            dest_filename = os.path.basename(parsed_url.path)
+            dest_path = os.path.join(dest_dir, dest_filename)
+            self.log.debug('Downloading %s', url)
+
+            if not os.path.exists(dest_dir):
+                os.makedirs(dest_dir)
+
+            req_session = get_retrying_requests_session()
+            request = req_session.get(url, stream=True,
+                                      verify=not insecure)
+            request.raise_for_status()
+
+            with open(dest_path, 'wb') as f:
+                for chunk in request.iter_content(chunk_size=DEFAULT_DOWNLOAD_BLOCK_SIZE):
+                    f.write(chunk)
+
+            self.log.debug('Download finished: %s', dest_path)
+            downloaded_sources.append(dest_path)
+
+        return downloaded_sources
+
+    def get_image_build_id(self):
+        nvr_build_id = None
+        if self.koji_build_nvr:
+            # strict means this raises a koji.GenericError informing no matching build was found in
+            # case the build does not exist
+            nvr_build = self.session.getBuild(self.koji_build_nvr, strict=True)
+            nvr_build_id = nvr_build['build_id']
+
+        if self.koji_build_id and nvr_build_id and (self.koji_build_id != nvr_build_id):
+            err_msg = (
+                'koji_build_id {} does not match koji_build_nvr {} with id {}. '
+                'When specifying both an id and an nvr, they should point to the same image build'
+                .format(self.koji_build_id, self.koji_build_nvr, nvr_build_id)
+                )
+            raise ValueError(err_msg)
+        return self.koji_build_id or nvr_build_id
+
+    def get_srpm_urls(self, build_id):
+        """
+        Fetch SRPM download URLs for each image generated by a build
+
+        :param root_url: str, Koji root storage URL
+        :return: list, strings with URLs pointing to SRPM files
+        """
+        srpm_urls = []
+        archives = self.session.listArchives(build_id, type='image')
+        for archive in archives:
+            rpms = self.session.listRPMs(imageID=archive['id'])
+            for rpm in rpms:
+                rpm_hdr = self.session.getRPMHeaders(rpm['id'], headers=['SOURCERPM'])
+                rpm_build = self.session.getBuild(rpm['build_id'])
+                path_info = get_koji_path_info(self.workflow, self.koji_fallback)
+                srpm_url = '/'.join([path_info.build(rpm_build), 'src', rpm_hdr['SOURCERPM']])
+                # TODO: this should actually use a sigkey from the signing_intent, as:
+                # srpm_url = '/'.join([path_info.build(rpm_build), 'data', 'signed', sigkey, 'src', rpm_hdr['SOURCERPM']])
+                srpm_urls.append(srpm_url)
```

Comment:
Will we end up with duplicates when multiple architectures installed RPMs from the same SRPM?